morf-core/src/main/java/org/alfasoftware/morf/xml/Version2to3TranformingReader.java

```diff
@@ -0,0 +1,205 @@
+   * should have the transform applied.
+   * <p>
+   * This is designed to match the known output format of
+   * {@link XmlDataSetConsumer} which previously produced invalid XML. It is
+   * deliberately brittle. There is no need for a more intelligent XML parser
+   * here.
+   * </p>
+   *
+   * @param bufferedReader The input stream in a buffered reader
+   * @return true if the transform should be applied. (because it's format 2)
+   */
+  static boolean shouldApplyTransform(BufferedReader bufferedReader) {
+    try {
+      bufferedReader.mark(100); // arbitrary read-ahead limit of 100 - that's enough to get the info we want
+      try {
+        {
+          String line = bufferedReader.readLine();
+          // the first line is probably the xml declaration
+          boolean isXmlDeclaration = line.startsWith("<?xml") && line.endsWith("?>");
+          if (!isXmlDeclaration) {
+            return false;
+          }
+        }
+        {
+          String line = bufferedReader.readLine();
+          // the next line is probably the table element
+          boolean isTableElement = line.startsWith("<table version");
+          if (!isTableElement) {
+            return false;
+          }
+
+          // Apply the transform if the version number is 2 or 1
+          return line.contains("version=\"2\"") || line.contains("version=\"1\"");
+        }
+
+      } finally {
+        bufferedReader.reset();
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+
+  /**
+   * @see java.io.Reader#read(char[], int, int)
+   */
+  @Override
+  public int read(char[] cbuf, int off, int len) throws IOException {
+    // We need to transform &#0; into \0...
+    int charsRead;
+
+    // if there's no temporary buffer from a previous call, read from the main source
+    if (temporary.length == 0) {
+      // This is the common path
+      charsRead = delegateReader.read(cbuf, off, len);
+    } else {
+      // there is a temporary buffer from a previous match, use that
+      if (temporary.length > len) {
+        // The temporary buffer is too big to fit in the buffer that's been supplied. This is an edge case, but we need to deal with it.
+        // Copy out what we can, then create another temporary buffer for the remainder
+        System.arraycopy(temporary, 0, cbuf, off, len);
+        charsRead = len;
+        char[] newTemporary = new char[temporary.length-len];
+        System.arraycopy(temporary, len, newTemporary, 0, temporary.length-len);
+        temporary = newTemporary;
+      } else {
+        // copy the entire temporary buffer into the output
+        System.arraycopy(temporary, 0, cbuf, off, temporary.length);
+        charsRead = temporary.length;
+        temporary = new char[] {};
+      }
+    }
+
+    // now search for the string we're replacing
+    for (int idx=0; idx<charsRead; idx++) {
+      if (cbuf[off+idx] == nullRefChars[0]) { // look for the ampersand

```

Comment:
Can we try and extract some of these expressions into descriptive methods - would make it a bit more readable.

Reply:
Done.